File: archive/2020/solutions/TurtleTribulation.py
from collections import deque

class Point:
    def __init__(self, x = 0, y = 0):
        self.x = x
        self.y = y

    def __lt__(self, other):
        return self.y > other.y

INF = 1000000000

n, z, k = input().split()
n = int(n)
z = int(z)
k = float(k)

p = []
for i in range(n):
    x, y = [int(i) for i in input().split()]
    p.append(Point(x, y))
p.sort()

dp = [INF] * n * (z + 1)
dp[p[0].x] = 0
for i in range(1, n):
    dq = deque()
    ptr1 = ptr2 = 0
    for j in range(z + 1):
        l = max(j - int((p[i-1].y - p[i].y) / k), 0)
        r = min(j + int((p[i-1].y - p[i].y) / k), z)
        while ptr2 < l:
            if dq[0][1] == ptr2:
                dq.popleft()
            ptr2 += 1
        while ptr1 <= r:
            while dq and dp[(i-1) * (z+1) + ptr1] <= dq[-1][0]:
                dq.pop()
            dq.append((dp[(i-1) * (z+1) + ptr1], ptr1))
            ptr1 += 1
        dp[i * (z+1) + j] = dq[0][0] + abs(j - p[i].x)

print(dp[(n-1) * (z+1) + p[n-1].x])

File: archive/2020/solutions/TurtleTribulation.java
import java.util.*;
import java.io.*;

// By Ayush Varshney
public class TurtleTribulation {

	public static void main(String[] args) throws IOException {
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt(), z = sc.nextInt();
		double k = sc.nextDouble();
		Point[] in = new Point[n];
		for (int i = 0; i < n; i++)
			in[i] = new Point(sc.nextInt() - 1, sc.nextInt() - 1);
		Arrays.sort(in);
		int[][] dp = new int[n][z];
		Arrays.fill(dp[0], 99999999);
		dp[0][in[0].a] = 0;
		for (int i = 1; i < n; i++) {
			Deque<Point> d = new ArrayDeque<>();
			for (int j = 0, l = 0; j < z; j++) {
				while (l <= Math.min(j + (int) ((in[i].b - in[i - 1].b) / k), z - 1)) {
					while (!d.isEmpty() && d.peekLast().a >= dp[i - 1][l])
						d.pollLast();
					d.addLast(new Point(dp[i - 1][l], l++));
				}
				if (d.peekFirst().b == Math.max(j - (int) ((in[i].b - in[i - 1].b) / k), 0) - 1)
					d.pollFirst();
				dp[i][j] = Math.abs(in[i].a - j) + d.peekFirst().a;
			}
		}
		System.out.println(dp[n - 1][in[n - 1].a]);
		sc.close();
	}

	static class Point implements Comparable<Point> {
		int a;
		int b;

		Point(int a, int b) {
			this.a = a;
			this.b = b;
		}

		@Override
		public int compareTo(Point o) {
			return this.b - o.b;
		}
	}
}